### PR TITLE
Adds Controls Config file as a JSON

### PR DESCRIPTION
**Pull request by Viv-0**
Adds a controls config file as a JSON.
Tested against Windows 10 with Controller and Keyboard
Supports deadzones for joysticks/triggers
Please send feedback if you have concerns

### CODE REVIEW

File: Source/Assets.cs [hunk 1]
```diff
@@ -176,11 +177,22 @@ public static void Load()
...
+#pragma warning disable CS8600
+            JsonNode doc = JsonNode.Parse(File.ReadAllText(controlsPath)); 
+			if (doc != null)
+				Controls.HandleCustomControls(doc);
 		}
 
-		// wait for tasks to finish
-		{
-			foreach (var task in tasks)
+#pragma warning restore CS8600
```
Comment (steviegt6) [inline Source/Assets.cs:192]: Why pragma instead of `!`?

File: Source/Controls.cs [hunk 2]
```diff
@@ -82,31 +85,187 @@ public static void Consume()
...
+        ("Jump", Jump ),
+        ("Dash", Dash ),
+        ("Climb", Climb ),
+        ("Menu Confirm", Confirm ),
+        ("Menu Back", Cancel ),
+        ("Pause", Pause )
+    };
+
+#pragma warning disable CS8600
+#pragma warning disable CS8604
```
Comment (steviegt6) [inline Source/Controls.cs:131]: Why not just handle nullability?

File: Source/Controls.cs [hunk 3]
```diff
@@ -82,31 +85,187 @@ public static void Consume()
...
+    // i probably should have handled it a bit better but :(, im just hoping that this isn't the slowest code in the game. - Viv
+    public static void HandleCustomControls(JsonNode root)
```
Comment (steviegt6) [inline Source/Controls.cs:133]: Would it not be better to deserialize to a type-checked object?

Comment (steviegt6) [reply Source/Controls.cs:133]: This existing code is extremely volatile and will cause problems if even minor issues are present.

**Review by steviegt6 — CHANGES_REQUESTED** [review]
Also lots of indentation inconsistencies.

**Comment by NoelFB** [comment]
Hey thanks for writing this all out! This is a great start, but I also agree with steviegt6 on a few issues:
 - Using null checks/handling null is better than using pragma's to ignore them. Ideally (aside from external libs) we wouldn't ignore any warnings.
 - I think creating a typed class to automatically serialize into might be worth doing. It might also make parsing the end data a lot easier. Since we're trimming the exe it will need a JsonSerializerContext but that's fairly trivial to do (see how [Save.cs does it](https://github.com/ExOK/Celeste64/blob/main/Source/Save.cs#L161))

**Comment by bitten2up** [comment]
just tested this out, seems like the pause menu isn't working at least for keyboard

a quick look through the changes I'm not quite sure what the issue is

I have tried both a debug build and a release build on linux and both can't open it

**Comment by Viv-0** [comment]
Regarding this, I think in order to use a native Save structure it would require rewriting things in terms of solely buttons, similar to the structure within Celeste itself, i.e.:
`"MoveUp": { "Controller": ["LeftStickUp", "DPadUp"], "Keyboard": ["Up"] }`

I'm going to rewrite it terms of that but I'm not sure whether or not I should then include Deadzones for things like Joystick/Triggers

**Comment by steviegt6** [comment]
You can write your own (de)serializers.

**Comment by NoelFB** [comment]
I ended up doing an implementation getting it working, based on the ideas here! Appreciate the first pass that was done, thank you!

It's still missing Mouse to aim, but I think that could be added as a separate PR (and probably part of Foster somehow).

**Pull request closed without merging**